src/backend/Pg.cpp

```diff
@@ -61,90 +61,291 @@ PgResult::msg() const
+    }
+
+    if (PQstatus(conn_.get()) != CONNECTION_OK)
+    {
+        std::stringstream ss;
+        ss << "bad connection" << std::to_string(PQstatus(conn_.get()));
+        char* err = PQerrorMessage(conn_.get());
+        if (err)
+            ss << ": " << err;
+        else
+            ss << '.';
+        throw std::runtime_error(ss.str());
+    }
+}
+
+inline void
+Pg::flush(boost::asio::yield_context& yield)
+{
+    auto& socket = getSocket();
+
+    // non-blocking connection requires manually flushing write.
+    int flushed;
+    do
+    {
+        flushed = PQflush(conn_.get());
+        if (flushed == 1)
+        {
+            socket.async_wait(
+                boost::asio::ip::tcp::socket::wait_write, yield);
+        }
+        else if (flushed == -1)
+        {
+            std::stringstream ss;
+            ss << "error flushing query " << PQerrorMessage(conn_.get());
+            throw std::runtime_error(ss.str());
+        }
+    }
+    while (flushed);
+}
+
+inline PgResult
+Pg::waitForStatus(boost::asio::yield_context& yield, ExecStatusType expected)
+{
+    auto& socket = getSocket();
+
+    PgResult ret;
+    while (true)
+    {
+        if (PQisBusy(conn_.get()))
+        {
+            socket.async_wait(boost::asio::ip::tcp::socket::wait_read,
+                yield);
+        }
+
+        if (!PQconsumeInput(conn_.get()))
+        {
+            std::stringstream ss;
+            ss << "query consume input error: "
+                << PQerrorMessage(conn_.get());
+            throw std::runtime_error(ss.str());
+        }
+
+        if (PQisBusy(conn_.get()))
+            continue;
+        
+        pg_result_type res{PQgetResult(conn_.get()),
+                            [](PGresult *result)
+                            { PQclear(result); }};
+        
+        if (!res)
+            break;
+
+        auto status = PQresultStatus(res.get());
+        ret = PgResult(std::move(res));
+        
+        if (status == expected)
+            break;
+    }
+
+    return ret;
+}
+
+inline boost::asio::ip::tcp::socket&
+Pg::getSocket()
+{
+    if (!socket_ || socketNum_ != PQsocket(conn_.get()))
+    {
+        socketNum_ = PQsocket(conn_.get());
+        socket_ = std::make_unique<boost::asio::ip::tcp::socket>(strand_,
+            boost::asio::ip::tcp::v4(), PQsocket(conn_.get()));
+    }
+
+    return *socket_;
 }
 
 PgResult
-Pg::query(char const* command, std::size_t nParams, char const* const* values)
+Pg::query(
+    char const* command,
+    std::size_t nParams, 
+    char const* const* values,
+    boost::asio::yield_context& yield)
 {
-    // The result object must be freed using the libpq API PQclear() call.
-    pg_result_type ret{nullptr, [](PGresult* result) { PQclear(result); }};
+    pg_result_type ret {nullptr, [](PGresult* result){ PQclear(result); }};
     // Connect then submit query.
-    while (true)
+    try
     {
+        connect(yield);
+
+        int sent;
+        if (nParams)
```

Comment:
I'm pretty sure PQsendQueryParams() with 0 nParams and null values is the same as PQsendQuery(). It's handled the same as PQexec & PQexecParams (the synchronous version): https://www.postgresql.org/docs/12/libpq-exec.html

So there should be no reason for calling the separate functions, nor for keeping the sent variable since the status can be checked for a single function call.